include/dds/core/xtypes/detail/MemberType.hpp

```diff
@@ -18,12 +18,195 @@
+    }
+
+    template<typename AnnoIter>
+    void annotation(
+            AnnoIter begin,
+            AnnoIter end)
+    {
+        ann_.reserve(ann_.size() + ( end - begin) ) ;
+        for (auto it = begin ; it != end ; ++it)
+        {
+            ann_.emplace_back(*it) ;
+        }
+    }
+
+    void annotation(
+            xtypes::Annotation &ann)
+    {
+        ann_.push_back(ann)  ;
+    }
+
+    const std::string &name()const noexcept
+    {
+        return name_ ;
+    }
+
+    const xtypes::DynamicType &dt() const noexcept
+    {
+        return dt_ ;
+    }
+
+    const std::vector< xtypes::Annotation >& annotation()
+    {
+        return ann_ ;
+    }
+
+    void remove_annotation(
+            const xtypes::Annotation &a)
+    {
+        auto rem = std::find_if(
+                    ann_.begin(),
+                    ann_.end(),
+                    [&]( xtypes::Annotation &b)
+                        {return b.akind() == a.akind();} ) ;
+        if ( rem != ann_.end() )
+        {
+            ann_.erase(rem) ;
+        }
+    }
+
+
+    std::vector<xtypes::Annotation>::iterator annIt(
+            AnnotationKind &ann)
+    {
+        return std::find_if(
+                        ann_.begin(),
+                        ann_.end(),
+                        [&]( xtypes::Annotation &a)
+                            { return (a.akind() == ann) ;} ) ;
+    }
+
+    bool findAnnotation(
+            AnnotationKind &ann)
+    {
+        auto found = annIt(ann) ;
+        return found != ann_.end() ;
+    }
+
+    bool is_optional()
+    {
+        AnnotationKind  a = AnnotationKind_def::Type::OPTIONAL_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    bool is_shared()
+    {
+        AnnotationKind a = AnnotationKind_def::Type::SHARED_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    bool is_key()
+    {
+        AnnotationKind a = AnnotationKind_def::Type::KEY_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    bool is_must_understand()
+    {
+        AnnotationKind a = AnnotationKind_def::Type::MUST_UNDERSTAND_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    bool is_bitset()
+    {
+        AnnotationKind a = AnnotationKind_def::Type::BITSET_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    bool has_bitbound()
+    {
+        AnnotationKind a = AnnotationKind_def::Type::BITSETBOUND_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    uint32_t get_bitbound()
+    {
+        if( false == has_bitbound() )
+        {
+            throw IllegalOperationError("No Bitsetbound Annotation found") ;
+        }
+        AnnotationKind a = AnnotationKind_def::Type::BITSETBOUND_ANNOTATION_TYPE ;
+        annIt(a)->bound() ;
+    }
+
+    bool has_id()
+    {
+        AnnotationKind a = AnnotationKind_def::Type::ID_ANNOTATION_TYPE ;
+        return findAnnotation(a) ;
+    }
+
+    uint32_t get_id()
+    {
+        if( false == has_id() )
```

Comment:
I think that is no need to check it before if you use later the `annIt`. You can check if the returned vector is empty and avoid iterate two times.

Reply:
the problem with that has_id() was due to a doubt I had with casting over iterators. I solved it with a simple refactoring